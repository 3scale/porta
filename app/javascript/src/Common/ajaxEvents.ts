/**
 * Set up AJAX event global listeners for Rails UJS (remote forms and links). They need to be
 * declared with jQuery so that they can be overridden by local listeners (e.stopPropagation).
 * Otherwise they will run after any other listeners, possibly breaking the UX.
 * Example:
 *   document.body.addEventListener('ajax:error', (event) => {
 *     // Do stuff to local component
 *
 *     // Prevent global listener to be called
 *     event.stopPropagation()
 *   })
 *
 * Origin: app/assets/javascripts/ajax_events.js
 */

import $ from 'jquery'

import type { AJAXErrorEvent } from 'Types/rails-ujs'

/**
 * rails-ujs ajax events.
 */
$(document)
  .on('ajax:before', (...args) => {
    console.log('ajax:before on', args)
    window.ThreeScale.showSpinner()
  })
  .on('ajax:beforeSend', (...args) => { console.log('ajax:beforeSend', args) })
  .on('ajax:send', (...args) => { console.log('ajax:send on', args) })
  .on('ajax:stopped', (...args) => { console.log('ajax:stopped', args) })
  .on('ajax:success', (...args) => { console.log('ajax:success', args) })
  .on('ajax:error', (event) => {
    const [response, status] = (event.originalEvent as AJAXErrorEvent<unknown>).detail
    console.error('ajax:error', response, status)
  })
  .on('ajax:complete', (...args) => {
    console.log('ajax:complete', args)
    window.ThreeScale.hideSpinner()
  })

/**
 * window.$.ajax events, triggered by Colorbox.
 */
window.$(document)
  .on('ajaxStart', (...args) => { console.log('ajaxStart', args) })
  .on('ajaxSend', (...args) => { console.log('ajaxSend', args) })
  .on('ajaxSuccess', (...args) => { console.log('ajaxSuccess', args) })
  .on('ajaxComplete', (...args) => { console.log('ajaxComplete', args) })
  .on('ajaxError', (...args) => { console.log('ajaxError', args) })

/**
 * Handle errors in formtastic forms rendered inside a colorbox. The error template is passed a
 * responseText inside the XHR response.
 *
 * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
 * TODO: add a colorbox specific selector to separate this with non-legacy implementations.
 */
$(document).on('ajax:error', 'form[data-remote]:not(.pf-c-form)', (event) => {
  const [,, xhr] = (event.originalEvent as AJAXErrorEvent<HTMLDocument>).detail

  // If EventTarget is replaceWith'ed immediately, ajax:success and ajax:complete are not called and the spinner will stay
  setTimeout(() => {
    window.$.colorbox({ html: xhr.responseText })
  })
  event.stopPropagation()
})
