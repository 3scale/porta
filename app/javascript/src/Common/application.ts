import type { JQueryXHR } from 'Types/jquery/v1'

const jQuery1 = window.$

export default function (): void {
  // disable links with 'data-disabled' attribute and display alert instead
  // delegation on body fires before rails.js. FIXME: this is not a valid guard. If "data-disabled"
  // removed manually from the DOM, the request will still go through. The link should be disabled
  // (not clickable) and the server should return a flash error.
  $('body').on('click', 'a[data-disabled]', ({ currentTarget }) => {
    alert((currentTarget as HTMLAnchorElement).dataset.disabled)
    return false
  })

  /**
   * This seems to be used only in app/views/master/providers/plans/_widget.html.slim:23
   * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
   * TODO: add a more specific selector to separate this form from all other "remote" ones.
   */
  $(document).on('submit', 'form.colorbox[data-remote]', ({ currentTarget }) => {
    // MUST USE RAILS-JQUERY since ajax:complete is triggered by rails (do not use $ from node_modules).
    const $form = jQuery1(currentTarget as HTMLFormElement)
    $form.on('ajax:complete', (_event, xhr: JQueryXHR) => {
      jQuery1.colorbox({
        open: true,
        html: xhr.responseText,
        width: $form.data('width') as string | undefined,
        maxWidth: '85%',
        maxHeight: '90%'
      })
    })
  })

  // TODO: replace .fancybox with .colorbox
  // This link will load its content into a colorbox modal
  $(document).on('click', 'a.fancybox, a.colorbox', ({ currentTarget }) => {
    jQuery1(currentTarget as HTMLAnchorElement).colorbox({ open: true })
    return false
  })

  // TODO: replace .fancybox with .colorbox
  // This is used in some modals with a "Cancel" button.
  $(document).on('click', '.fancybox-close', () => {
    jQuery1.colorbox.close()
    return false
  })

  /**
   * Handle errors in formtastic forms rendered inside a colorbox. The error template is passed a
   * responseText inside the XHR response. MUST USE RAILS-JQUERY since ajax:error is triggered by
   * rails (do not use $ from node_modules).
   *
   * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
   * TODO: add a colorbox specific selector to separate this with non-legacy implementations.
   */
  jQuery1(document).on('ajax:error', 'form:not(.pf-c-form)', (_event, xhr: JQueryXHR, status) => {
    if (status === 'error') {
      jQuery1.colorbox({ html: xhr.responseText })
      return false
    }
  })
}
