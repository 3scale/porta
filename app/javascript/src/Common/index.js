// @flow

export * from './components/SelectModal'
