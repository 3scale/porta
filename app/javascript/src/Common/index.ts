export { CompactListCard } from './components/CompactListCard'
export { FancySelect } from './components/FancySelect'
export { MicroPagination } from './components/MicroPagination'
export { NoMatchFound } from './components/NoMatchFound'
export { TableModal } from './components/TableModal'
export { ToolbarSearch } from './components/ToolbarSearch'
export { Select } from './components/Select'
export { SelectWithModal } from './components/SelectWithModal'
export { Spinner } from './components/Spinner'
export { SystemNamePopover, SystemNamePopoverWrapper } from './components/SystemNamePopover'
export { UserDefinedField } from './components/UserDefinedField'
