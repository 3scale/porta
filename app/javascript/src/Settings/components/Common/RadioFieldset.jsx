// @flow

import * as React from 'react'
import { useState } from 'react'
import { FormFieldset, FormLegend } from 'Form'
import { Radio } from '@patternfly/react-core'
import type { FieldGroupProps, FieldCatalogProps } from 'Settings/types'

type CheckEvent = SyntheticEvent<HTMLButtonElement>

type Props = FieldGroupProps & FieldCatalogProps

const useSelectedOnChange = (value, onChange) => (
  typeof onChange === 'function'
    ? [value, onChange]
    : useState(value).map(x => typeof x === 'function' ? (_c, e: CheckEvent) => x(e.currentTarget.value) : x)
)

const RadioFieldset = ({
  children,
  legend,
  name,
  value,
  catalog,
  onChange,
  ...props
}: Props): React.Node => {
  const [selectedOnChange, setSelectedOnChange] = useSelectedOnChange(value, onChange)
  return (
    // $FlowIgnore[cannot-spread-inexact]
    <FormFieldset id={`fieldset-${name}`} {...props} >
      <FormLegend>{legend}</FormLegend>
      {Object.keys(catalog).map(key => (
        <Radio
          key={key}
          value={key}
          isChecked={selectedOnChange === key}
          name={`service[${name}]`}
          onChange={setSelectedOnChange}
          label={catalog[key]}
          id={`service_method_${name}_${key}`}
        />
      ))}
      {children}
    </FormFieldset>
  )
}

export {
  RadioFieldset
}
