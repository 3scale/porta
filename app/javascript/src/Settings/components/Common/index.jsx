// @flow

export * from 'Settings/components/Common/FormCollection'
export * from 'Settings/components/Common/FormFieldset'
export * from 'Settings/components/Common/FormLegend'
export * from 'Settings/components/Common/SelectGroup'
export * from 'Settings/components/Common/TextInputGroup'
export * from 'Settings/components/Common/RadioFieldset'
export * from 'Settings/components/Common/TypeItemCombo'
