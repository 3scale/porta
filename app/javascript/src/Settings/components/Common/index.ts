export { FormCollection } from 'Settings/components/Common/FormCollection'
export { FormFieldset } from 'Settings/components/Common/FormFieldset'
export { FormLegend } from 'Settings/components/Common/FormLegend'
export { SelectGroup } from 'Settings/components/Common/SelectGroup'
export { TextInputGroup } from 'Settings/components/Common/TextInputGroup'
export { RadioFieldset } from 'Settings/components/Common/RadioFieldset'
export { TypeItemCombo } from 'Settings/components/Common/TypeItemCombo'
