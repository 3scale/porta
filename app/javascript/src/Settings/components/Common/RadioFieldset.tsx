/* eslint-disable react/jsx-props-no-spreading */
import { useState } from 'react'
import { Radio } from '@patternfly/react-core'
import { FormFieldset } from 'Settings/components/Common/FormFieldset'
import { FormLegend } from 'Settings/components/Common/FormLegend'

import type { FieldCatalogProps, FieldGroupProps } from 'Settings/types'

type CheckEvent = React.SyntheticEvent<HTMLButtonElement>

type Props = FieldGroupProps & FieldCatalogProps

// FIXME: WTF is this? Refactor and make it properly
const useSelectedOnChange = (value: string, onChange: undefined | ((value: string, event: React.SyntheticEvent<HTMLButtonElement>) => void)) => (
  typeof onChange === 'function'
    ? [value, onChange]
    // eslint-disable-next-line react/hook-use-state
    : useState(value).map(x => typeof x === 'function' ? (_c: any, e: CheckEvent) => x(e.currentTarget.value) : x) as any
)

const RadioFieldset: React.FunctionComponent<Props> = ({
  children,
  legend,
  name,
  value,
  catalog,
  onChange,
  ...props
}) => {
  const [selectedOnChange, setSelectedOnChange] = useSelectedOnChange(value as string, onChange)
  return (
    <FormFieldset id={`fieldset-${name}`} {...props} >
      <FormLegend>{legend}</FormLegend>
      {Object.keys(catalog).map(key => (
        <Radio
          key={key}
          id={`service_method_${name}_${key}`}
          isChecked={selectedOnChange === key}
          label={catalog[key]}
          name={`service[${name}]`}
          value={key}
          onChange={setSelectedOnChange}
        />
      ))}
      {children}
    </FormFieldset>
  )
}

export { RadioFieldset, Props }
