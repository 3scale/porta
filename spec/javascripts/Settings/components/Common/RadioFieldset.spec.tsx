import { shallow } from 'enzyme'
import { RadioFieldset } from 'Settings/components/Common/RadioFieldset'

import type { Props } from 'Settings/components/Common/RadioFieldset'

it('should render correctly', () => {
  const props: Props = {
    name: 'foo_fighters',
    value: 'pat',
    label: 'label',
    legend: 'Foo Fighters',
    checked: undefined,
    hint: undefined,
    placeholder: undefined,
    defaultValue: undefined,
    readOnly: undefined,
    inputType: undefined,
    isDefaultValue: undefined,
    onChange: undefined,
    catalog: {
      dave: 'Grohl',
      pat: 'Smear'
    }
  }

  const view = shallow(<RadioFieldset {...props} ><span>My Hero</span></RadioFieldset>)
  expect(view).toMatchSnapshot()
})
