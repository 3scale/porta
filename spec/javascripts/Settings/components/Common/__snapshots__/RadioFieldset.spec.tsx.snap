// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<FormFieldset
  id="fieldset-foo_fighters"
  label="label"
>
  <FormLegend>
    Foo Fighters
  </FormLegend>
  <Radio
    className=""
    id="service_method_foo_fighters_dave"
    isChecked={false}
    isDisabled={false}
    isValid={true}
    key="dave"
    label="Grohl"
    name="service[foo_fighters]"
    onChange={[Function]}
    value="dave"
  />
  <Radio
    className=""
    id="service_method_foo_fighters_pat"
    isChecked={true}
    isDisabled={false}
    isValid={true}
    key="pat"
    label="Smear"
    name="service[foo_fighters]"
    onChange={[Function]}
    value="pat"
  />
  <span>
    My Hero
  </span>
</FormFieldset>
`;
